src/plots/cartesian/axes.js

```diff
@@ -1610,14 +1610,41 @@ axes.drawOne = function(gd, ax, opts) {
     var subplotsWithAx = axes.getSubplots(gd, ax);
 
     var vals = ax._vals = axes.calcTicks(ax);
-    // We remove zero lines, grid lines, and inside ticks if they're within 1px of the end
-    // The key case here is removing zero lines when the axis bound is zero
-    var valsClipped = ax._valsClipped = axes.clipEnds(ax, vals);
 
     if(!ax.visible) return;
 
     var transFn = axes.makeTransFn(ax);
 
+    // We remove zero lines, grid lines, and inside ticks if they're within 1px of the end
+    // The key case here is removing zero lines when the axis bound is zero
+    var valsClipped;
+    var tickVals;
+    var gridVals;
+
+    if(ax.tickson === 'boundaries') {
+        // draw ticks and grid lines 1/2 a 'category' to the left,
+        // add one item at axis tail
+        var valsBoundaries = vals.map(function(d) {
+            var d2 = Lib.extendFlat({}, d);
+            d2.x -= 0.5;
+            return d2;
+        });
+        // not used for labels; no need to worry about the other tickTextObj keys
+        var d2 = Lib.extendFlat({}, vals[vals.length - 1]);
+        d2.x += 0.5;
```

Comment:
-0.5 for the other ticks is great, but +0.5 may need altering in case `dtick !== 1` (perhaps `+= dtick - 0.5`?). Have you experimented with this case? see eg https://github.com/plotly/plotly.js/issues/2601#issuecomment-386312022

Reply:
Ah good eye -> https://github.com/plotly/plotly.js/pull/3275/commits/6d0a3fc926a423061b954611def71abae858e73b